### ISSUE TITLE
categorical cross entropy goes to nan pretty easily

### ISSUE THREAD

**Issue by Vict0rSch** [fchollet/keras#1429]
Hi, here is my piece of code (standalone, you can try). On the last 5 times I tried, the loss went to `nan` before the 20th epoch. I just updated Keras and checked : in `objectives.py` `epsilon` is at: 

``` python
if theano.config.floatX == 'float64':
    epsilon = 1.0e-9
else:
    epsilon = 1.0e-7
```

code:

``` python
import time
import numpy as np
from matplotlib import pyplot as plt
from keras.utils import np_utils
import keras.callbacks as cb
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import RMSprop
from keras.datasets import mnist


class LossHistory(cb.Callback):
    def on_train_begin(self, logs={}):
        self.losses = []

    def on_batch_end(self, batch, logs={}):
        batch_loss = logs.get('loss')
        self.losses.append(batch_loss)


def load_data():
    print 'Loading data...'
    (X_train, y_train), (X_test, y_test) = mnist.load_data()

    X_train = X_train.astype('float32')
    X_test = X_test.astype('float32')

    X_train /= 255
    X_test /= 255

    y_train = np_utils.to_categorical(y_train, 10)
    y_test = np_utils.to_categorical(y_test, 10)

    X_train = np.reshape(X_train, (60000, 784))
    X_test = np.reshape(X_test, (10000, 784))

    print 'Data loaded.'
    return [X_train, X_test, y_train, y_test]


def init_model():
    start_time = time.time()
    print 'Compiling Model ... '
    model = Sequential()
    model.add(Dense(500, input_dim=784))
    model.add(Activation('relu'))
    model.add(Dropout(0.4))
    model.add(Dense(300))
    model.add(Activation('relu'))
    model.add(Dropout(0.4))
    model.add(Dense(10))
    model.add(Activation('softmax'))

    rms = RMSprop()
    model.compile(loss='categorical_crossentropy', optimizer=rms)
    print 'Model compield in {0} seconds'.format(time.time() - start_time)
    return model


def run_network(data=None, model=None, epochs=20, batch=256):
    try:
        start_time = time.time()
        if data is None:
            X_train, X_test, y_train, y_test = load_data()
        else:
            X_train, X_test, y_train, y_test = data

        if model is None:
            model = init_model()

        history = LossHistory()

        print 'Training model...'
        model.fit(X_train, y_train, nb_epoch=epochs, batch_size=batch,
                  callbacks=[history], show_accuracy=True,
                  validation_data=(X_test, y_test), verbose=2)

        print "Training duration : {0}".format(time.time() - start_time)
        score = model.evaluate(X_test, y_test, batch_size=16,
                               show_accuracy=True)

        print "Network's test score [loss, accuracy]: {0}".format(score)
        return model, history.losses
    except KeyboardInterrupt:
        print ' KeyboardInterrupt'
        return model, history.losses
```

**Comment by talentlei** [comment]
I have the same issue whenever I try to train GRU or LSTM with categorical_crossentropy. How can I solve it?

**Comment by svirg** [comment]
I also encountered the same issue
Setting epsilon=1e-05, and floatx=float64 in keras config file didn't seem to help. However everything gone fine when I switched theano optimizer mode from 'fast_compile' to 'fast_run' in .theanorc

**Comment by nouiz** [comment]
Do you use nvcc.fastmath=True? This is bad as it can generate nan in many
cases. This happen more frequently with optimizer=fast_compile.

If that don't help, you can try optimizer=stabilize, it include
stabilization optimization, but much less then fast_run, so it should
compile faster.

**Comment by 5olegas** [comment]
You can try to use another optimizator, Adam for example. Do not use RMSprop with categorical_crossentropy loss function.

**Comment by KingRando95** [comment]
Get Outlook for Android<https://aka.ms/ghei36>

________________________________
